hfile.c

```diff
@@ -506,12 +518,13 @@ static off_t fd_seek(hFILE *fpv, off_t offset, int whence)
 
 static int fd_flush(hFILE *fpv)
 {
-    hFILE_fd *fp = (hFILE_fd *) fpv;
-    int ret;
+    int ret = 0;
     do {
 #ifdef HAVE_FDATASYNC
+        hFILE_fd *fp = (hFILE_fd *) fpv;
         ret = fdatasync(fp->fd);
-#else
+#elif defined(HAVE_FSYNC)
+        hFILE_fd *fp = (hFILE_fd *) fpv;
         ret = fsync(fp->fd);
```

Comment:
I mean code like jmarshall/htslib@39146e27e146eba21d9243e8151682d0c33ce91d to ensure `fsync()` is always available (but probably moved to _hts_os.h_ or so).

Alternatively that code as a function in _hfile.c_ or just inline in `fd_flush()`, likely controlled by an explicit platform `HAVE_` macro like [these ones](https://github.com/samtools/htslib/blob/dc45f188e5c0bd296b9087ad6119722ebc48aca4/hfile.c#L457-L465).